### PR TITLE
ICU-22626 Fix leakage when 2 '=' in PluralRules

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22626
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/i18n/plurrule.cpp [hunk 1]
```diff
@@ -656,6 +656,10 @@ PluralRuleParser::parse(const UnicodeString& ruleData, PluralRules *prules, UErr
         case tEqual:
             {
                 U_ASSERT(curAndConstraint != nullptr);
```
Comment (roubert) [inline icu4c/source/i18n/plurrule.cpp:658]: It's illogical to have first this assertion, which indicates that this is something that shouldn't be possible, and then right after that have some code for handling it if it happens anyway. If this is something that actually can happen (as the fuzzer result suggests) then the assertion should be deleted (as it then isn't actually true).

If I follow the code correctly, this `PluralRuleParser::parse()` function is called from `PluralRules::createRules()` which is public API, meaning that a user can pass in any random string and that makes me think that all these assertions in this function are highly suspicious.

I would recommend you to widen the scope of ICU-22626 from just patching this one particular issue found by the fuzzer to investigating all the assertions in this function and replace all of those that can actually happen with proper error handling code.

File: icu4c/source/i18n/plurrule.cpp [hunk 2]
```diff
@@ -905,6 +909,10 @@ AndConstraint::AndConstraint(const AndConstraint& other) {
         if (U_FAILURE(fInternalStatus)) {
             return;
         }
+        if (this->rangeList != nullptr) {
```
Comment (roubert) [inline icu4c/source/i18n/plurrule.cpp:912]: This doesn't make any sense to me. This code here is in the constructor for `class AndConstraint` and I can't see any code before this line that assigns anything to `this->rangeList` so the code in this if-statement will never be executed. Is that so, or what am I missing?

**Review by roubert — CHANGES_REQUESTED** [review]

File: icu4c/source/i18n/plurrule.cpp (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/plurrule.cpp:658]: The assertion is on 
"curAndConstraint != nullptr", the if return is on 
"curAndConstraint->rangeList != nullptr" NOT "curAndConstraint != nullptr"
two different things.

File: icu4c/source/i18n/plurrule.cpp (hunk 2 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/plurrule.cpp:912]: you are right, I didn't realize it is in the constructor code.

**Comment by FrankYFTang** [comment]
PTAL

File: icu4c/source/i18n/plurrule.cpp (hunk 1 above)
Comment (roubert) [reply icu4c/source/i18n/plurrule.cpp:658]: Oops, now when I read it again it looks fine, sorry for the noise.

**Review by roubert — APPROVED** [review]
LGTM

**Pull request merged**